### PR TITLE
added optimized CleverNotifier

### PR DESCRIPTION
**Pull request by escamoteur**
I think I could improve the performance a bit.

### CODE REVIEW

**Comment by knaeckeKami** [comment]
did you remove the getX and LinkedListValueNotifier benchmarks on purpose?

**Comment by escamoteur** [comment]
I only removed them from the maps to get a easier to read and faster Benchmark. Do you think we still need them?

**Comment by knaeckeKami** [comment]
nah i think it's fine.

File: lib/notifiers/thomas2.dart [hunk 1]
```diff
@@ -1,217 +1,90 @@
...
-      }
-      return true;
-    }());
-    return true;
-  }
+class Thomas2ChangeNotifier {
+  List<_Listener> _listeners = List<_Listener>();
+  int _notificationCallStackDepth = 0;
+  int _removedListeners = 0;
+  int _listenerLengh; // This is the length on the first not recursive call
```
Comment (letsar) [inline lib/notifiers/thomas2.dart:18]: I think you wanted to wrote _listenerLength instead.

File: lib/notifiers/thomas2.dart [hunk 2]
```diff
@@ -1,217 +1,90 @@
...
-          },
-        ));
+    if (_listeners.isEmpty) {
+      return;
+    }
+    _notificationCallStackDepth++;
+
+    if (_listeners != null) {
+      _listenerLengh ??= _listeners.length;
+      for (int i = 0; i < _listenerLengh; i++) {
```
Comment (letsar) [inline lib/notifiers/thomas2.dart:55]: By doing this, we prevent listeners added during the first iteration to be called in the second iteration. We would want to remove this instance field and use a variable instead.

**Review by letsar — COMMENTED** [review]
We probably want to remove _listenerLengh because it prevents newly listeners to be called during a second iteration during the first notifyListener.

File: lib/notifiers/thomas2.dart (hunk 1 above)
Comment (escamoteur) [reply lib/notifiers/thomas2.dart:18]: yep

File: lib/notifiers/thomas2.dart (hunk 2 above)
Comment (escamoteur) [reply lib/notifiers/thomas2.dart:55]: actually I did this for exactly this reason. I'm not sure if it is a good idea to have a recursive call calling the newly added, but I honestly think we should ass an assert that forbids recursive calls

Comment (letsar) [reply lib/notifiers/thomas2.dart:55]: But by doing this, we could break existing code, I'm not sure we should change the current behavior.

**Comment by escamoteur** [comment]
I also think we should make the List nonnullable with nullable elemts and ad a disposed flag instead of setting the List null on dispose.
Actually I don't know why a Notifier should no longer be used after a dispose

File: lib/notifiers/thomas2.dart (hunk 2 above)
Comment (escamoteur) [reply lib/notifiers/thomas2.dart:55]: I thought the old Notifier didn't support recursion at all?

Comment (letsar) [reply lib/notifiers/thomas2.dart:55]: I'm not sure, but since the list was cloned, I think it was possible to do it.

Comment (escamoteur) [reply lib/notifiers/thomas2.dart:55]: Can we check this? I think @knaeckeKami or you said they did not support recursion. If so we should check that we have the same behaviour

**Comment by escamoteur** [comment]
can we switch this benchmark to null safety? @knaeckeKami

**Comment by knaeckeKami** [comment]
yeah, PR's welcome, if not I can migrate today evening or maybe tomorrow. 
Would be interesting to see if the performance is noticeably better when we use sound null safety, be we'd have to drop the dependency on get and I'd have to migrate barbecue (because the compiler optimizations for null safety are only enabled when all dependencies are migrated too)

**Comment by escamoteur** [comment]
you don't have to opt in to full safety with all packages

**Comment by knaeckeKami** [comment]
yeah you can use null safety also with non null safe packages, but afaik then you don't get the additional compiler optimizations. 
See

> One caveat: We only guarantee soundness in Dart programs that are fully null safe. Dart supports programs that contain a mixture of newer null safe code and older legacy code. In these mixed-version programs, null reference errors may still occur. In a mixed-version program, you get all of the static safety benefits in the portions that are null safe, but you don’t get full runtime soundness until the entire application is null safe.
[Source](https://dart.dev/null-safety/understanding-null-safety)

**Comment by knaeckeKami** [comment]
Migration is to nullsafety is not that easy. dart migrate throws an exception and I don't want to migrate by hand, so we'll have to wait until the Dart team fixes this:
See https://github.com/dart-lang/sdk/issues/44349

In the meantime, you could extract your ValueNotifier in another dart project with nullsafety enabled (could also be in this repo, we could use a monorepo structure)

**Comment by escamoteur** [comment]
It seems we are pretty good in finding the edge cases of the migration tool :-) GetIt alone turned up several new bugs

**Comment by knaeckeKami** [comment]
Yeah, although I didn't have any issues with my other packages 😅

**Comment by knaeckeKami** [comment]
migrated to nullsafety on the "nullsafety3" branch. Mostly just applied to `dart migrate` tool, with some small corrections. Did not change any logic and did not apply any suggestions of @mraleph from the thread yesterday yet.

**Pull request merged**